File: reservation-frontend/src/app/core/services/salle.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Salle } from '../models/salle/salle';
import { environment } from '../../../environments/environment.development';


@Injectable({
  providedIn: 'root'
})
export class SalleService {

  // URL de base correspondant à @RequestMapping("/api/salles")
  private readonly apiUrl = `${environment.apiUrl}/api/salles`;

  constructor(private http: HttpClient) {}

  /**
   * GET /api/salles
   * Récupérer toutes les salles
   */
  getAllSalles(): Observable<Salle[]> {
    return this.http.get<Salle[]>(this.apiUrl);
  }

  /**
   * GET /api/salles/{id}
   * Récupérer une salle par ID
   */
  getSalleById(id: number): Observable<Salle> {
    return this.http.get<Salle>(`${this.apiUrl}/${id}`);
  }

  /**
   * POST /api/salles
   * Ajouter une nouvelle salle
   */
  createSalle(salle: Salle): Observable<Salle> {
    return this.http.post<Salle>(this.apiUrl, salle);
  }

  /**
   * PUT /api/salles/{id}
   * Modifier une salle existante
   */
  updateSalle(id: number, salle: Salle): Observable<Salle> {
    return this.http.put<Salle>(`${this.apiUrl}/${id}`, salle);
  }

  /**
   * DELETE /api/salles/{id}
   * Supprimer une salle
   */
  deleteSalle(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  /**
   * GET /api/salles/ville/{ville}
   * Rechercher les salles par ville
   */
  findByVille(ville: string): Observable<Salle[]> {
    return this.http.get<Salle[]>(`${this.apiUrl}/ville/${ville}`);
  }

  /**
   * GET /api/salles/capacite/{capacite}
   * Rechercher les salles par capacité minimale
   */
  findByCapacite(capacite: number): Observable<Salle[]> {
    return this.http.get<Salle[]>(`${this.apiUrl}/capacite/${capacite}`);
  }

  /**
   * GET /api/salles/recherche?ville=...&capacite=...
   * Recherche multicritère (ville et capacité)
   */
  rechercherSalleDisponible(ville: string, capacite: number): Observable<Salle[]> {
    const params = new HttpParams()
      .set('ville', ville)
      .set('capacite', capacite.toString());

    return this.http.get<Salle[]>(`${this.apiUrl}/recherche`, { params });
  }
}

File: Reservation/src/main/java/net/travel/reservation/services/SalleService.java
package net.travel.reservation.services;

import lombok.RequiredArgsConstructor;
import net.travel.reservation.entites.Salle;
import net.travel.reservation.repositories.SalleRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class SalleService {

    private final SalleRepository salleRepository;

    // Récupérer toutes les salles
    public List<Salle> getAllSalles() {
        return salleRepository.findAll();
    }

    // Récupérer une salle par ID
    public Salle getSalleById(Long id) {
        return salleRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Salle introuvable avec l'ID : " + id));
    }

    // Ajouter une salle
    @Transactional
    public Salle createSalle(Salle salle) {
        if (salle.getNom() != null && salleRepository.existsByNom(salle.getNom())) {
            throw new RuntimeException("Une salle avec le nom '" + salle.getNom() + "' existe déjà");
        }
        return salleRepository.save(salle);
    }

    // Modifier une salle
    @Transactional
    public Salle updateSalle(Long id, Salle salleRequest) {
        Salle salle = getSalleById(id);

        // Vérification si le nom a changé et s'il est déjà utilisé
        if (salleRequest.getNom() != null &&
                !salleRequest.getNom().equals(salle.getNom()) &&
                salleRepository.existsByNom(salleRequest.getNom())) {
            throw new RuntimeException("Une autre salle porte déjà le nom : " + salleRequest.getNom());
        }

        salle.setNom(salleRequest.getNom());
        salle.setCapaciteMax(salleRequest.getCapaciteMax());
        salle.setDescription(salleRequest.getDescription());
        salle.setAdresse(salleRequest.getAdresse());
        salle.setVille(salleRequest.getVille());
        salle.setTelephone(salleRequest.getTelephone());
        salle.setEmail(salleRequest.getEmail());

        return salleRepository.save(salle);
    }

    // Supprimer une salle
    @Transactional
    public void deleteSalle(Long id) {
        Salle salle = getSalleById(id);
        salleRepository.delete(salle);
    }

    // Recherche par ville
    public List<Salle> findByVille(String ville) {
        return salleRepository.findByVille(ville);
    }

    // Recherche par capacité minimale
    public List<Salle> findByCapacite(Integer capacite) {
        return salleRepository.findByCapaciteMaxGreaterThanEqual(capacite);
    }

    // Recherche de salles correspondant aux critères
    public List<Salle> rechercherSalleDisponible(String ville, Integer capacite) {
        return salleRepository.findByVilleAndCapaciteMaxGreaterThanEqual(ville, capacite);
    }
}